File: src/service/config/service_config.zig
// src/config/service_config.zig

const std = @import("std");

pub const ServiceConfig = struct {
    /// Service name (e.g. "pricing", "order-gateway").
    service_name: []const u8,

    /// Control ring buffer capacity. Power of 2.
    control_buffer_size: u32 = 65_536,

    /// Application messages ring buffer capacity. Power of 2.
    messages_buffer_size: u32 = 1_048_576,

    /// Whether the ring buffer uses kernel-level blocking when full.
    blocking_mode: bool = false,

    /// Service heartbeat timeout in milliseconds.
    heartbeat_timeout_ms: u32 = 10_000,

    /// Idle strategy for the service's message consumer thread.
    idle_strategy_name: []const u8 = "backoff",

    /// Whether this service participates in per-service leader election.
    leader_election_enabled: bool = false,

    /// Storage path override (defaults to broker's storage path).
    storage_path: ?[]const u8 = null,

    /// Group name override (defaults to broker's group name).
    group_name: ?[]const u8 = null,
};

File: bindings/java/src/main/java/io/ringloom/service/ServiceConfig.java
package io.ringloom.service;

import java.util.Objects;

/**
 * Configuration for starting a {@link RingloomService}.
 *
 * @param serviceName logical service name to register
 * @param storagePath shared-memory storage path, or {@code null} for the default
 * @param group broker group name, or {@code null} for the default
 * @param brokerNodeId broker node id, or {@code 0} for the default
 * @param blockingMode whether native operations should use blocking behavior where supported
 * @param heartbeatTimeoutMillis heartbeat timeout, or {@code 0} for the default
 * @param controlBufferLength control ring buffer size, or {@code 0} for the default
 * @param messagesBufferLength message ring buffer size, or {@code 0} for the default
 * @param leaderElectionEnabled whether this service participates in leader election
 */
public record ServiceConfig(
    String serviceName,
    String storagePath,
    String group,
    short brokerNodeId,
    boolean blockingMode,
    int heartbeatTimeoutMillis,
    long controlBufferLength,
    long messagesBufferLength,
    boolean leaderElectionEnabled
) {
    public static final String DEFAULT_STORAGE_PATH = "/dev/shm";
    public static final String DEFAULT_GROUP = "default";
    public static final short DEFAULT_BROKER_NODE_ID = 1;
    public static final int DEFAULT_HEARTBEAT_TIMEOUT_MILLIS = 10_000;
    public static final long DEFAULT_CONTROL_BUFFER_LENGTH = 65_536L;
    public static final long DEFAULT_MESSAGES_BUFFER_LENGTH = 1_048_576L;

    /**
     * Validates the service name and applies defaults for optional zero or null fields.
     */
    public ServiceConfig {
        serviceName = Objects.requireNonNull(serviceName, "serviceName");
        if (serviceName.isEmpty()) {
            throw new IllegalArgumentException("serviceName must not be empty");
        }

        storagePath = storagePath == null ? DEFAULT_STORAGE_PATH : storagePath;
        group = group == null ? DEFAULT_GROUP : group;
        brokerNodeId = brokerNodeId == 0 ? DEFAULT_BROKER_NODE_ID : brokerNodeId;
        heartbeatTimeoutMillis = heartbeatTimeoutMillis == 0
            ? DEFAULT_HEARTBEAT_TIMEOUT_MILLIS
            : heartbeatTimeoutMillis;
        controlBufferLength = controlBufferLength == 0
            ? DEFAULT_CONTROL_BUFFER_LENGTH
            : controlBufferLength;
        messagesBufferLength = messagesBufferLength == 0
            ? DEFAULT_MESSAGES_BUFFER_LENGTH
            : messagesBufferLength;
    }

    /**
     * Creates a configuration using default values for all optional fields.
     *
     * @param serviceName logical service name to register
     * @return service configuration with defaults applied
     */
    public static ServiceConfig of(String serviceName) {
        return new ServiceConfig(serviceName, null, null, (short) 0, false, 0, 0, 0, false);
    }
}
